
---
title: "Processing seurat object"
author: "Ivo Borkus"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, message = FALSE, warning = FALSE,
    fig.align = "center", fig.width = 10, fig.height = 10
)
knitr::render_markdown()
```

```{css, echo = F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}

pre {
  max-height: 300px;
  overflow-y: auto;
}
```


```{r loading libraries, include = FALSE}
library(Seurat)
library(dplyr)
library(ggplot2)
library(here)
library(tidyr)
library(patchwork)
library(qs2)
source(here("Brain-Met_V2/functions.R"))
# source(here("External_data/Fastq_processing/metadata_functions.R"))
```
# Setting paths and names

```{r}
# sub_dir <- "external"
exc_path <- here("Glioma_Project/01_data/Excel")
rmd_path <- here("Glioma_Project/02_knits/figs")
obj_path <- here("Glioma_Project/01_data")
fig_path <- here("Glioma_Project/04_Figs")



object_name <- "cnag_DG_immune.rds"
reduction_umap <- "umap"
reduction_pca <- "pca"
n_PCS <- 50

# CR_path <- here("Brain-Met_V2/data/Cellranger")
minUMI <- 500
maxUMI <- Inf
minfeat <- 500
maxfeat <- Inf
maxmt <- 20
sub_dir <- NULL
batch_col <- "tumour_abbreviated"
sample_col <- "sample_name"
batch_cols <- c(
    "tumour_full", "tumour_abbreviated", "grade", "Cell_Type", "Cell_Subtype", "sample_name", "IDH_status",
    "FACS_sorting", "tier_correct", "cell_het_paper_id", "storage_condition", "prior_treatment", "gender", "Tumour_Margin",
    "Primary_or_Recurrent", "Previous_Surgery", "Radiotherapy", "Chemotherapy", "MGMT_Methylation"
)
continous_col <- c("nCount_RNA", "nFeature_RNA", "percent.mt", "age_at_surgery_correct", "nuclear_fraction")
```



```{r plotting some things}
feature_vec_overall <- c(
    "PTPRC", "CD3E", "CD3D", "CD4", "CD8A", "CD8B", "TOP2A", "MKI67", # T and prolif
    "CD1C", "HLA-DQB1", # DC
    "S100AB", "TREM2", "LGALS3", # Monocyte macrophage
    "TMEM119", # Microglial
    "MOG" # Oligos
)
feature_vec_T <- c("PTPRC", "CD8B", "CD4", "LAG3", "CX3CR1", "CCR7", "SELL", "NCAM1", "NKG7", "KLRB1", "FOXP3", "TOP2A", "MKI67", "PDCD1", "FCGR3A", "MAG")

feature_vec_chosen <- feature_vec_overall
```

```{r read}
seurat_obj <- readRDS(here(obj_path, object_name))
for (list_col in colnames(seurat_obj@meta.data)[lapply(seurat_obj@meta.data, class) == "list"]) {
    list_vec <- seurat_obj@meta.data[[list_col]]
    list_vec <- sapply(list_vec, function(x) if (length(x) == 0) NA else x)
    seurat_obj@meta.data[[paste0(list_col, "_correct")]] <- unlist(list_vec)
    print(paste0(list_col, "_correct"))
}

pattern <- "^([A-Z]+)_(?:\\d{4}_)?(\\d{3}).*"
sub(pattern, "\\1_\\2", head(seurat_obj@meta.data[[sample_col]]))

seurat_obj@meta.data$Sample <- sub(pattern, "\\1_\\2", seurat_obj@meta.data[[sample_col]])

seurat_obj@meta.data$Batch <- seurat_obj@meta.data[[batch_col]]
seurat_obj@meta.data$percent_mt <- seurat_obj@meta.data$percent.mt
for (column in colnames(seurat_obj@meta.data)) {
    print(column)
    seurat_obj@meta.data[[column]] <- ifelse(is.na(seurat_obj@meta.data[[column]]), "Not available", seurat_obj@meta.data[[column]])
    seurat_obj@meta.data[[column]] <- ifelse(seurat_obj@meta.data[[column]] == "NA", "Not available", seurat_obj@meta.data[[column]])
}
```


# Loading seurat object

```{r}
c(minUMI, maxUMI, minfeat, maxfeat, maxmt) %>%
    as.data.frame() %>%
    t() %>%
    as.data.frame() -> y
names(y) <- c("minUMI", "maxUMI", "minfeat", "maxfeat", "maxmt")


minUMI <- ifelse(is.na(minUMI), min(seurat_obj@meta.data$nCount_RNA), minUMI)
maxUMI <- ifelse(is.na(maxUMI), max(seurat_obj@meta.data$nCount_RNA), maxUMI)
minfeat <- ifelse(is.na(minfeat), min(seurat_obj@meta.data$nFeature_RNA), minfeat)
maxfeat <- ifelse(is.na(maxfeat), max(seurat_obj@meta.data$nFeature_RNA), maxfeat)
maxmt <- ifelse(is.na(maxmt), max(seurat_obj@meta.data$percent_mt), maxmt)

x <- .adjust_settings(seurat_obj, minUMI, maxUMI, minfeat, maxfeat, maxmt)
minUMI <- x[1]
maxUMI <- x[2]
minfeat <- x[3]
maxfeat <- x[4]
x %>%
    as.data.frame() %>%
    t() %>%
    as.data.frame() -> x2
names(x2) <- c("minUMI", "maxUMI", "minfeat", "maxfeat", "maxmt")
setting_df <- rbind(y, x2)
setting_df <- cbind(Settings = c("Input paramaters", "New paramater values"), setting_df)
gt::gt(setting_df) %>% gt::tab_header(gt::md("**Settings after adjustments**"))
```

```{r, echo = F, message = F, fig.width = 14, fig.height = 14}
# shared_scale <- scale_fill_manual(name = "Time point", labels = c("Post", "Pre"), values = c("#FF474C", "skyblue"))
gg_co <- .cell_count(seurat_obj, filling_var = "Batch")
gg_nc <- .ncount_plot(seurat_obj, fill_variable = "Batch", min_umi = minUMI, max_umi = maxUMI) # 500 & 10000
gg_nf <- .nfeat_plot(seurat_obj, fill_variable = "Batch", min_count = minfeat, max_count = maxfeat) # 500 & 10000
gg_mt <- .nmt_plot(seurat_obj, low_mt = maxmt, fill_variable = "Batch")

(gg_co + gg_mt + gg_nc + gg_nf) +
    plot_layout(guides = "collect") &
    # shared_scale &
    theme(axis.title.y = element_blank())
```



```{r, echo = F}
umi_v_mt_lines <- FeatureScatter(seurat_obj,
    feature1 = "nCount_RNA", feature2 = "percent_mt", group.by = "Batch",
    pt.size = 1
)
umi_v_genes_lines <- FeatureScatter(seurat_obj,
    feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "Batch", pt.size = 1
)
```

```{r, fig.width = 14, fig.height = 14}
umi_v_mt_lines + umi_v_genes_lines
```

```{r, fig.width = 14, fig.height = 14}
p <- list()
for (name in unique(seurat_obj@meta.data$Sample)) {
    seurat_obj_meta_filter <- seurat_obj@meta.data %>% filter(Sample == name)
    Ncells <- dim(seurat_obj_meta_filter)[1]
    print(Ncells)
    plot <- ggplot(
        data = seurat_obj_meta_filter,
        mapping = aes(
            x = nCount_RNA, y = nFeature_RNA,
            color = percent_mt,
            size = percent_mt
        )
    ) +
        geom_point() +
        scale_color_gradient(limits = c(0, 100)) +
        scale_size(limits = c(0, 100)) +
        scale_x_log10(limits = c(min(seurat_obj@meta.data$nCount_RNA), max(seurat_obj@meta.data$nCount_RNA))) +
        scale_y_log10(limits = c(min(seurat_obj@meta.data$nFeature_RNA), max(seurat_obj@meta.data$nFeature_RNA))) +
        labs(
            x = "UMI count", y = "Unique genes", title = paste0(name, " - ", Ncells)
        ) +
        theme(
            panel.border = element_blank(), panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
            axis.text = element_text(size = 12), axis.title = element_text(size = 20), title = element_text(size = 15), axis.text.x = element_text(angle = 45, hjust = 1)
        ) +
        geom_hline(yintercept = minfeat, color = "red") +
        geom_hline(yintercept = maxfeat, color = "red") +
        geom_vline(xintercept = minUMI, color = "red") +
        geom_vline(xintercept = maxUMI, color = "red")

    p[[name]] <- plot
}
```


```{r tabsetting}
a <- 0
n <- 6
plot_list <- list()
for (i in seq(from = 1, to = length(p), by = n)) {
    a <- a + 1
    if ((i + (n - 1)) > length(p)) {
        plot <- wrap_plots(p[i:length(p)], guides = "collect", axis_titles = "collect")
    } else {
        plot <- wrap_plots(p[i:(i + (n - 1))], guides = "collect", axis_titles = "collect")
    }
    plot_list[[paste0("Samples: ", i, "-", (i + (n - 1)))]] <- plot
}
```

## individual plots {.tabset}


```{r , echo = F, results='asis', fig.width = 14, fig.height = 14}
.tabsetter(plot_list)
```

```{r, eval = T,echo = F, message= F}
total_cells <- dim(seurat_obj)[2]
seurat_obj_sub <- subset(
    seurat_obj,
    nCount_RNA >= minUMI &
        nCount_RNA <= maxUMI &
        nFeature_RNA >= minfeat &
        nFeature_RNA <= maxfeat &
        percent_mt <= maxmt
)

subset_cells <- dim(seurat_obj_sub)[2]
data_removed <- 100 - ((subset_cells / total_cells) * 100)

###

pre <- seurat_obj$Sample %>%
    table()
post <- seurat_obj_sub$Sample %>%
    table()
df <- as.data.frame(((pre - post) / pre) * 100)
df$Before <- pre
df$After <- post
names(df) <- c("Sample", "Filtered %", "Before", "After")
df_arranged <- df %>% arrange(desc(`Filtered %`))

# temp <- list()

# for (sample_name in df_arranged$Sample) {
#     df_temp <- seurat_obj@meta.data %>% filter(Sample == sample_name)
#     Batch_ID <- unique(df_temp$Batch)
#     Tissue_ID <- unique(df_temp$Tissue)
#     Time_point <- unique(df_temp$TP)

#     temp[[sample_name]] <- c(Batch_ID, Tissue_ID, Time_point)
# }
# temp_T <- temp %>%
#     as.data.frame() %>%
#     t()
# new_df <- cbind(df_arranged, temp_T)
# row.names(new_df) <- NULL


# names(new_df) <- c("Sample", "Filtered %", "Before", "After", "Batch", "Tissue", "Time point")
```

# Filtering outcome overall

```{r}
cat(paste0("Total amount of cells: ", total_cells, "\n"))
cat(paste0("Percentage of data removed: ", round(data_removed, 2), "%", "\n"))
cat(paste0("Amount of cells removed: ", total_cells - subset_cells, "\n"))
cat(paste0("Cells remaining: ", subset_cells))
```

# Filtering per sample

```{r}
library(gt)
gt::gt(df_arranged) %>%
    gt::tab_header(
        title = md("**Filtering results with strict settings**"),
        subtitle = (paste0("Settings: Min count - ", minUMI, "; Min,Max feat - (", minfeat, ",", maxfeat, "); max Mt - ", maxmt))
    ) %>%
    gt::cols_label(
        Sample = md("**Sample**"),
        `Filtered %` = md("**Filtered %**"),
        Before = md("Before"),
        After = md("After"),
        # Batch = md("Batch"),
        # Tissue = md("Tissue"),
        # `Time point` = md("Time point")
    )
```



```{r}
sessionInfo()
```

```{r old functions, echo = F, eval = F}
library(here)
rmarkdown::render("Glioma_Project/Scripting/QC_seurat.rmd", output_dir = "Glioma_Project/02_knits", output_file = paste0(Sys.Date(), "Seurat_QC_Glioma.html"))

# ggsave(filename = here(fig_path, paste0("Cell_count_per_sample_", Sys.Date(), "_.pdf")), dpi = 300)
# ggsave(filename = here(fig_path, paste0("Cell_count_per_sample_", Sys.Date(), "_.png")), dpi = 500)


# gg_nc <- .ncount_plot(seurat_obj, fill_variable = "TP")
# ggsave(filename = here(fig_path, paste0("ncount_plot_", Sys.Date(), "_.pdf")), dpi = 300)
# ggsave(filename = here(fig_path, paste0("ncount_plot_", Sys.Date(), "_.png")), dpi = 500)


# gg_nf <- .nfeat_plot(seurat_obj, fill_variable = "TP")
# ggsave(filename = here(fig_path, paste0("nfeat_plot_", Sys.Date(), "_.pdf")), dpi = 300)
# ggsave(filename = here(fig_path, paste0("nfeat_plot_", Sys.Date(), "_.png")), dpi = 500)

# gg_mt <- .nmt_plot(seurat_obj, low_mt = NULL, fill_variable = "TP")
# ggsave(filename = here(fig_path, paste0("nmt_plot_", Sys.Date(), "_.pdf")), dpi = 300)
# ggsave(filename = here(fig_path, paste0("nmt_plot_", Sys.Date(), "_.png")), dpi = 500)
```

