
---
title: "Processing seurat object"
author: "Ivo Borkus"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, message = FALSE, warning = FALSE,
    fig.align = "center", fig.width = 10, fig.height = 10
)
knitr::render_markdown()
```

```{css, echo = F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}

pre {
  max-height: 300px;
  overflow-y: auto;
}
```


```{r loading libraries, include = FALSE}
library(Seurat)
library(dplyr)
library(ggplot2)
library(here)
library(tidyr)
library(patchwork)
library(qs2)
source(here("Brain-Met_V2/functions.R"))
# source(here("External_data/Fastq_processing/metadata_functions.R"))
```
# Setting paths and names

```{r}
exc_path <- here("Glioma_Project/01_data/Excel")
rmd_path <- here("Glioma_Project/02_knits/figs")
obj_path <- here("Glioma_Project/01_data")
fig_path <- here("Glioma_Project/04_Figs")

object_name <- "cnag_DG_immune.rds"
reduction_name <- "Inhouse_processed"
Processed_name <- "Processed"
sample_col <- "sample_name"
batch_cols <- c(
    "tumour_full", "tumour_abbreviated", "grade", "Cell_Type", "Cell_Subtype", "sample_name", "IDH_status",
    "FACS_sorting", "tier_correct", "cell_het_paper_id", "storage_condition", "prior_treatment", "gender", "Tumour_Margin",
    "Primary_or_Recurrent", "Previous_Surgery", "Radiotherapy", "Chemotherapy", "MGMT_Methylation"
)
continous_cols <- c("nCount_RNA", "nFeature_RNA", "percent.mt", "age_at_surgery_correct", "nuclear_fraction")


# subsetting <- c("Cell_Type", c("Myeloid", "Dendritic cells", "Mast cells")) #  Subset-col, cells_of_interest
subsetting <- c(NULL, NULL) #  Subset-col, cells_of_interest

norm_meth_scale <- c(TRUE, "LogNormalize", 10000) # Perform log normalisatin, method, scale
vf <- c("vst", 2000, FALSE, FALSE) # VF method, amount, use old vf method, silence VDJgenes using scRep
scaling <- TRUE #
var.to.regress <- NULL # supply the var to be regressed if valid
harmony <- F # Run harmony or not
n_PCS <- 50 # Amount of Pcs to include
clustering <- c(TRUE, 4) # Perform clustering, Algorithm
sub_dir <- NULL


feature_vec_chosen <- feature_vec_overall
```


```{r read}
seurat_obj <- readRDS(here(obj_path, object_name))
for (list_col in colnames(seurat_obj@meta.data)[lapply(seurat_obj@meta.data, class) == "list"]) {
    list_vec <- seurat_obj@meta.data[[list_col]]
    list_vec <- sapply(list_vec, function(x) if (length(x) == 0) NA else x)
    seurat_obj@meta.data[[paste0(list_col, "_correct")]] <- unlist(list_vec)
    print(paste0(list_col, "_correct"))
}

pattern <- "^([A-Z]+)_(?:\\d{4}_)?(\\d{3}).*"

seurat_obj@meta.data$Sample <- sub(pattern, "\\1_\\2", seurat_obj@meta.data[[sample_col]])

seurat_obj@meta.data$Batch <- seurat_obj@meta.data[["Sample"]]
seurat_obj@meta.data$percent_mt <- seurat_obj@meta.data$percent.mt
for (column in colnames(seurat_obj@meta.data)) {
    print(column)
    seurat_obj@meta.data[[column]] <- ifelse(is.na(seurat_obj@meta.data[[column]]), "Not available", seurat_obj@meta.data[[column]])
    seurat_obj@meta.data[[column]] <- ifelse(seurat_obj@meta.data[[column]] == "NA", "Not available", seurat_obj@meta.data[[column]])
}
```
```{r processing seurat, eval = F, message = F, include = T}
# Requires a joined seurat object
seurat_obj <- UpdateSeuratObject(seurat_obj)
seurat_obj <- scCustomize::Convert_Assay(seurat_object = seurat_obj, convert_to = "V5")
# seurat_obj <- JoinLayers(seurat_obj)
seurat_obj <- .processing(
    obj = seurat_obj,
    reduction_name = reduction_name,
    Processed_name = Processed_name,
    subsetting = subsetting,
    norm_meth_scale = norm_meth_scale,
    vf = vf,
    scaling = scaling,
    var.to.regress = var.to.regress,
    harmony = harmony,
    n_PCS = n_PCS,
    clustering = clustering
)
```


```{R loading seurat}
seurat_obj_path <- ifelse(length(sub_dir) > 0, here(obj_path, sub_dir, paste0(Processed_name, ".qs2")), here(obj_path, paste0(Processed_name, ".qs2")))
seurat_obj <- qs_read(seurat_obj_path)
.size(seurat_obj)
```


```{r checking PCs}
reduction_pca <- paste0("pca_", reduction_name)
reduction_umap <- paste0("umap_", reduction_name)
seurat_obj@meta.data %>% colnames()
PCA_elbow <- ElbowPlot(seurat_obj, reduction = reduction_pca, ndims = 200)
pct <- seurat_obj[[reduction_pca]]@stdev / sum(seurat_obj[[reduction_pca]]@stdev) * 100

cumu <- cumsum(pct)

co1 <- which(cumu > 90 & pct < 5)[1]
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
pcs <- min(co1, co2)
plot_df <- data.frame(
    pct = pct,
    cumu = cumu,
    rank = 1:length(pct)
)
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pcs)) +
    geom_text() +
    geom_vline(xintercept = 90, color = "grey") +
    geom_hline(yintercept = min(pct[pct > 2]), color = "grey") +
    theme_bw()


print("Theoretical PC to use: ")
print(pcs)
```

```{r QC processing}
df_corr <- .regress_PCA(seurat_obj, reduction_name, harmony, n_PCS, c(batch_cols, continous_cols))

plot_list_batch <- plotting_PCA_regress(df_corr, limit_n = 1)
PCA_elbow <- ElbowPlot(seurat_obj, reduction = reduction_pca, ndims = 200)
```

```{r plotting, results = 'asis', echo = F}
.tabsetter(plot_list_batch, overall_name = "Batch plots", path = rmd_path)
```

```{r}
plots <- .Running_plots(seurat_obj,
    reduction_name = reduction_name, output_dir = fig_path,
    harmony = harmony, umap_feature_vec = feature_vec_chosen, batch_cols = batch_cols
)
```

# Results
```{R}
DimPlot(seurat_obj, reduction = reduction_umap, group.by = "Sample") & NoAxes()
DimPlot(seurat_obj, reduction = reduction_umap, group.by = "Cell_Type") & NoAxes()
DimPlot(seurat_obj, reduction = reduction_umap, group.by = "Cell_Subtype") & NoAxes()

FeaturePlot(seurat_obj, reduction = reduction_umap, features = c("nCount_RNA", "nFeature_RNA", "percent_mt")) & NoAxes()

library(writexl)
resolution <- "RNA_snn_res.0.05"
Idents(seurat_obj) <- resolution
DimPlot(seurat_obj, reduction = reduction_umap, group.by = resolution)

marker_list <- FindAllMarkers(
    object = seurat_obj,
    only.pos = F,
    min.pct = 0.25,
    logfc.threshold = 0.25
)
.excel_sheet(marker_list, exc_path, reduction_name)
```

## The top 5 DEX markers for each cluster
```{r}
top5_wide <- marker_list %>%
    group_by(cluster) %>%
    arrange(desc(avg_log2FC), p_val_adj, .by_group = TRUE) %>%
    slice_head(n = 5) %>%
    select(cluster, gene) %>%
    pivot_wider(names_from = cluster, values_from = gene)

top5_wide %>%
    knitr::kable(format = "markdown", caption = "Top 5 marker genes per cluster")
```

```{r session}
sessionInfo()
```

rmarkdown::render("Glioma_Project/Scripting/02_processing_external.rmd", output_dir = "Glioma_Project/02_knits", output_file = paste0(Sys.Date(), "Seurat_processing_external_GLioma.html")) 
