
---
title: "Processing seurat object"
author: "Ivo Borkus"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = T, message = FALSE, warning = FALSE,
    fig.align = "center", fig.width = 10, fig.height = 10
)
# knitr::render_markdown()
```

```{css, echo = F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}

pre {
  max-height: 300px;
  overflow-y: auto;
}
```


```{r loading libraries, include = FALSE}
library(Seurat)
library(dplyr)
library(ggplot2)
library(here)
library(tidyr)
library(patchwork)
library(UCell)
library(qs2)
source(here("Brain-Met_V2/functions.R"))
# source(here("External_data/Fastq_processing/metadata_functions.R"))
```

# Setting paths and names

Here we set the variables necessary to run my custom scripts. The rest of the script therefore should not be tweaked. 

However, as there is an interpretation stage, all code after Results is custom for each object.
```{r}
exc_path <- here("Glioma_Project/01_data/Excel")
rmd_path <- here("Glioma_Project/02_knits/figs")
obj_path <- here("Glioma_Project/01_data")
fig_path <- here("Glioma_Project/04_figs")

object_name <- "Harmony"
reduction_name <- "Myeloid_Harmony_2000"
Processed_name <- "Myeloid_Harmony_2000"
sample_col <- "sample_name"
batch_cols <- c(
    "tumour_full", "tumour_abbreviated", "grade", "Cell_Type", "Cell_Subtype", "sample_name", "IDH_status",
    "FACS_sorting", "tier_correct", "cell_het_paper_id", "storage_condition", "prior_treatment", "gender", "Tumour_Margin",
    "Primary_or_Recurrent", "Previous_Surgery", "Radiotherapy", "Chemotherapy", "MGMT_Methylation"
)
continous_cols <- c("nCount_RNA", "nFeature_RNA", "percent.mt", "age_at_surgery_correct", "nuclear_fraction")


# subsetting <- c("Cell_Type", c("Myeloid", "Dendritic cells", "Mast cells")) #  Subset-col, cells_of_interest
subsetting <- c("Ann_lvl_1", c("Microglia", "Macrophage", "Myeloid_Prolif", "Mast-cells", "pDCs")) #  Subset-col, cells_of_interest

norm_meth_scale <- c(F, "LogNormalize", 10000) # Perform log normalisatin, method, scale
vf <- c("vst", 2000, FALSE, FALSE) # VF method, amount, use old vf method, silence VDJgenes using scRep
scaling <- TRUE #
var.to.regress <- NULL # supply the var to be regressed if valid
harmony <- T # Run harmony or not
n_PCS <- 50 # Amount of Pcs to include
clustering <- c(TRUE, 4) # Perform clustering, Algorithm
sub_dir <- NULL
feature_vec_chosen <- feature_vec_overall
```

# Read in the object that is to be processed
```{r read}
seurat_obj_path <- ifelse(length(sub_dir) > 0, here(obj_path, sub_dir, paste0(object_name, ".qs2")), here(obj_path, paste0(object_name, ".qs2")))
seurat_obj <- qs_read(seurat_obj_path)
.size(seurat_obj)
```

# Processing object

```{r processing seurat, eval = F, message = F, include = T}
# Requires a joined seurat object of v5. Code bellow can be used to do this
# seurat_obj <- UpdateSeuratObject(seurat_obj)
# seurat_obj <- scCustomize::Convert_Assay(seurat_object = seurat_obj, convert_to = "V5")
# seurat_obj <- JoinLayers(seurat_obj)
seurat_obj <- .processing(
    obj = seurat_obj,
    reduction_name = reduction_name,
    Processed_name = Processed_name,
    subsetting = subsetting,
    norm_meth_scale = norm_meth_scale,
    vf = vf,
    scaling = scaling,
    var.to.regress = var.to.regress,
    harmony = harmony,
    n_PCS = n_PCS,
    clustering = clustering
)
```

# Loading of processed seurat

```{R loading seurat}
seurat_obj_path <- ifelse(length(sub_dir) > 0, here(obj_path, sub_dir, paste0(Processed_name, ".qs2")), here(obj_path, paste0(Processed_name, ".qs2")))
seurat_obj <- qs_read(seurat_obj_path)
.size(seurat_obj)
```

# Perform QC on the processed object

## Extracting variation of PCs

```{r, eval = !harmony, echo = !harmony}
reduction_pca <- paste0("pca_", reduction_name)
reduction_umap <- paste0("umap_", reduction_name)
pct <- seurat_obj[[reduction_pca]]@stdev / sum(seurat_obj[[reduction_pca]]@stdev) * 100
```

```{r, eval = harmony, echo = harmony}
reduction_pca <- paste0("harmony_pca.", reduction_name)
reduction_umap <- paste0("umap.harmony.", reduction_name)
emb <- seurat_obj[[reduction_pca]]@cell.embeddings # 212005 Ã— 200
variance <- apply(emb, 2, var)
pct <- variance / sum(variance) * 100
```

## Object info.

```{r checking columns}
seurat_obj@meta.data %>% colnames() # Possible columns to investigate
```

## Elbow plot

```{r checking PCS}
cumu <- cumsum(pct)

co1 <- which(cumu > 90 & pct < 5)[1]
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
pcs <- min(co1, co2)
plot_df <- data.frame(
    pct = pct,
    cumu = cumu,
    rank = 1:length(pct)
)
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > n_PCS)) +
    geom_text() +
    geom_vline(xintercept = 90, color = "grey") +
    geom_hline(yintercept = min(pct[pct > 2]), color = "grey") +
    theme_bw()
```


```{r, echo = F}
print("Theoretical PC to use by some reddit post: ")

print(pcs)
```

```{r QC processing,eval = !harmony, echo = !harmony}
df_corr <- .regress_PCA(seurat_obj, reduction_name, harmony, n_PCS, c(batch_cols, continous_cols))

plot_list_batch <- plotting_PCA_regress(df_corr, limit_n = 1)
PCA_elbow <- ElbowPlot(seurat_obj, reduction = reduction_pca, ndims = 200)
```

# Regression of PCs

```{r QC processing harmony,eval = harmony, echo = harmony}
features <- c(batch_cols, continous_cols)
df_corr_before <- .regress_PCA(seurat_obj, reduction_name, F, n_PCS, features)
df_corr_after <- .regress_PCA(seurat_obj, reduction_name, harmony, n_PCS, features)
df_corr_before$Type <- "Before"
df_corr_after$Type <- "After"

# Combine the data
df_plot <- bind_rows(df_corr_before, df_corr_after)
QC_values <- unique(df_plot$QC)

# Create an empty list to store plots
plot_list_batch <- list()

# Loop over each QC and create a plot
for (qc in QC_values) {
    plot_list_batch[[qc]] <- ggplot(
        df_plot %>% filter(QC == qc),
        aes(x = PCs, y = cor, color = Type, linetype = Type, group = Type)
    ) +
        geom_line() +
        ylim(c(0, 1)) +
        theme_classic() +
        theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1, size = 6)) +
        scale_linetype_manual(values = c("Before" = "dotted", "After" = "solid")) +
        scale_color_manual(values = c("Before" = "red", "After" = "green")) +
        labs(x = "PCs", y = "Correlation", title = paste("QC:", qc))
}

#####
# This plot is not used, but can be used to summarise all batch cols in one figure
plot <- ggplot(df_plot, aes(x = PCs, y = cor, group = Type, linetype = Type, color = Type)) +
    geom_line() +
    ylim(c(0, 1)) +
    facet_wrap(~QC) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1, size = 6)) +
    scale_linetype_manual(values = c("Before" = "dotted", "After" = "solid")) +
    scale_color_manual(values = c("Before" = "red", "After" = "green")) +
    labs(x = "PCs", y = "Correlation", title = "Original vs Batch-Corrected Correlations")
```

```{r plotting, results = 'asis', echo = F}
.tabsetter(plot_list_batch, overall_name = "Batch plots", path = rmd_path)
```

# Checking the reduction

```{r}
plots <- .Running_plots(seurat_obj,
    reduction_name = reduction_name, output_dir = fig_path,
    harmony = harmony, umap_feature_vec = feature_vec_chosen, batch_cols = batch_cols
)
```


```{r tabsetting, results = 'asis', echo = F}
.tabsetter(plots[[1]], overall_name = paste0("Processing plots: ", reduction_name), path = rmd_path)
.tabsetter(plots[[2]], overall_name = paste0("Processing plots: ", reduction_name), path = rmd_path)
```

# Plotting signatures
```{r plotting signatures, eval = T}
gene_list <- list(
    BD = c("LYZ", "EREG", "S100A8", "S100A9", "ITGAL", "ITGA4", "PTPRC", "FCGR3A", "CEBPD", "EGR2", "IRF4", "RUNX1", "STAT3", "CCL22", "CXCL2", "CD14", "S100A4", "S100A6", "FCN1", "VCAN", "CD36", "LGALS3", "MRC1", "CD163", "SELENOP", "SLC40A1", "CXCR4"),
    MG = c("P2RY12", "CX3CR1", "CSF1R", "AXL", "TMEM119", "CD84", "CSF3R", "ELMO1", "MEF2A", "BIN1", "SCIN", "FCGBP", "OLFML3", "NAV3", "SALL1", "SIGLEC8", "SLC1A3", "ITGAM", "TREM2", "PROS1", "SPARC", "PLCL1", "VIM", "APOE", "SMAD3", "P2RX7", "LYVE1", "MS4A7", "MRC1", "FOLR2", "CD163", "HLA-DRA", "TIMD4", "ADGRG1", "TGFBI", "CD74"),
    Mast = c("IL1RL1", "HDC", "TNFSF10", "FGL2", "IL17RB", "CCL2", "CSF1", "IL13"),
    pDCs = c("GZMB", "IGJ", "SERPINF1", "ITM2C", "IRF7", "TCF4", "BCL11A", "LILRA4", "MZB1", "DERL3", "IL3RA", "ZFAT", "NRP1", "CLEC4C", "IRF8", "SLC15A4", "BLNK"),
    DC_general = c("CD1C", "CLEC9A"),
    Prolif = c("UBE2C", "MKI67", "TOP2A"),
    Mg_homeostatic = c("PTPRC", "ITGAM", "P2RY12"),
    Mg_activated = c("CX3CR1", "GPR34", "FOXP2"),
    Mg_Reslike = c("PLCL1", "CEBPD", "VIM"),
    Mg_stress = c("HSPA1A", "HSPA1B", "SORCS2"),
    Mg_Phag = c("GPNMB", "RGCC", "PPARG"),
    Mg_Inflam = c("CCL4", "IL1B", "CCL3"),
    Mg_Inflam_ICAM1 = c("ICAM1", "RELB", "TNFAIP3"),
    Mg_INFg = c("IFNG", "IFIT2", "IFIT3", "STAT1"),
    BAM = c("F13A1", "LYVE1", "CD36"),
    BMD_Anti_Inflam = c("CD163", "SELENOP", "MRC1")
)
```

```{r, eval = T}
seurat_obj <- AddModuleScore_UCell(seurat_obj,
    features = gene_list
)

qs_save(seurat_obj, seurat_obj_path)
```

```{r, eval = T}
resolution_columns <- grep("^RNA_snn_res\\.",
    colnames(seurat_obj@meta.data),
    value = TRUE
)
.seurat_dimplot(seurat_obj, reduction = reduction_umap, group = resolution_columns, label = TRUE) & NoLegend()
.seurat_dimplot(seurat_obj, reduction = reduction_umap, group = "Ann_lvl_1", label = TRUE) & NoLegend()
FeaturePlot(seurat_obj, features = paste0(names(gene_list), "_UCell")[1:8], reduction = reduction_umap) & NoAxes()
FeaturePlot(seurat_obj, features = paste0(names(gene_list), "_UCell")[9:16], reduction = reduction_umap) & NoAxes()
```



# Results

## Choosing a reduction and extracting marker genes



```{r testing code, eval = T}
library(writexl)
resolution <- "RNA_snn_res.0.7"

Idents(seurat_obj) <- resolution
DimPlot(seurat_obj, reduction = reduction_umap, group.by = resolution, label = T) & NoLegend() & NoAxes()
ggsave(here(fig_path, paste0(resolution, "_", reduction_name, "_Dimplot.png")))
DimPlot(seurat_obj, reduction = reduction_umap, group.by = "Ann_lvl_1", label = T) & NoLegend() & NoAxes()
ggsave(here(fig_path, paste0("Ann_lvl_1", "_", reduction_name, "_Dimplot.png")))


marker_list <- FindAllMarkers(
    object = seurat_obj,
    only.pos = F,
    min.pct = 0.25,
    logfc.threshold = 0.25
)
table(Idents(seurat_obj)) # Extract rough proportions.
.excel_sheet(marker_list, exc_path, paste0(reduction_name, "_", resolution)) # This just puts them into an excel sheet :)
```

## The top 5 DEX markers for each cluster

```{r}
top5_wide <- marker_list %>%
    group_by(cluster) %>%
    arrange(desc(avg_log2FC), p_val_adj, .by_group = TRUE) %>%
    slice_head(n = 5) %>%
    select(cluster, gene) %>%
    pivot_wider(names_from = cluster, values_from = gene)
```

## adding details of the cluster IDs

```{r adding details of the cluster IDs, eval = F}
# new.cluster.ids <- c("Microglia", "T_Prolif", "pDCs", "Mast-cells", "Microglia", "Macrophage", "Microglia", "T-cells", "Microglia", "Macrophage", "Myeloid_Prolif", "Macrophage", "Tregs")
# new.cluster.ids %>% length()
Idents(seurat_obj) <- resolution
names(new.cluster.ids) <- levels(seurat_obj)
seurat_obj <- RenameIdents(seurat_obj, new.cluster.ids)
seurat_obj[["Ann_lvl_1"]] <- Idents(seurat_obj)
DimPlot(seurat_obj, reduction = reduction_umap, group.by = "Ann_lvl_1", label = T)
colnames(top5_wide) <- new.cluster.ids
## Saving the annotations in the same processed object
qs_save(seurat_obj, seurat_obj_path)
```

```{r show the table}
gt::gt(top5_wide) %>%
    gt::tab_header(
        title = "Top 5 Marker Genes per Cluster"
    )
```

## Show the marker gene scores 

```{r plotlist scores}
plot_list_score <- lapply(names(gene_list), function(score) {
    p1 <- VlnPlot(seurat_obj, features = paste0(score, "_UCell"))
})
names(plot_list_score) <- names(gene_list)
```



```{r plotting scpre violins, results = 'asis', echo = F}
.tabsetter(plot_list_score, overall_name = "Scoring violin plots", path = rmd_path)
```

# Plotting confounders

```{r confounders}
VlnPlot(seurat_obj, features = c(c("nCount_RNA", "nFeature_RNA", "percent_mt")))
```

# SessionInfo
```{r session}
sessionInfo()
```

rmarkdown::render("Glioma_Project/Scripting/05_subset_myeloid_harmony.rmd", output_dir = "Glioma_Project/02_knits",
 output_file = paste0(Sys.Date(), "05_Myeloid_Harmony.html"))
