---
title: "Current annotation check"
author: "Ivo Borkus"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = T, message = FALSE, warning = FALSE,
  fig.align = "center", fig.width = 10, fig.height = 10
)
knitr::render_markdown()
```

```{css, echo = F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}

pre {
  max-height: 300px;
  overflow-y: auto;
}
```
```{r loading libraries, include = FALSE}
library(Seurat)
library(dplyr)
library(ggplot2)
library(here)
library(tidyr)
library(patchwork)
library(UCell)
library(qs2)
library(future)
plan(sequential)
options(future.globals.maxSize = 100000 * 1024^2)
# source(here("Brain-Met_V2/functions.R"))
source(here('Glioma_Project/Scripting/utils.R'))
source(here('Glioma_Project/Scripting/functions.R'))
# source(here("External_data/Fastq_processing/metadata_functions.R"))
```

`

# Setting paths and names

Here we set the variables necessary to run my custom scripts. The rest of the script therefore should not be tweaked. 

However, as there is an interpretation stage, all code after Results is custom for each object.
```{r}
exc_path <- here("Glioma_Project/01_data/Excel")
rmd_path <- here("Glioma_Project/02_knits/figs")
obj_path <- here("Glioma_Project/01_data")
fig_path <- here("Glioma_Project/04_figs")

object_name <- "Processed"
reduction_name <- "Inhouse_Harmony"
Processed_name <- "Harmony"
sample_col <- "sample_name"
batch_cols <- c(
  "tumour_full", "tumour_abbreviated", "grade", "Cell_Type", "Cell_Subtype", "sample_name", "IDH_status",
  "FACS_sorting", "tier_correct", "cell_het_paper_id", "storage_condition", "prior_treatment", "gender", "Tumour_Margin",
  "Primary_or_Recurrent", "Previous_Surgery", "Radiotherapy", "Chemotherapy", "MGMT_Methylation"
)
continous_cols <- c("nCount_RNA", "nFeature_RNA", "percent.mt", "age_at_surgery_correct", "nuclear_fraction")


# subsetting <- c("Cell_Type", c("Myeloid", "Dendritic cells", "Mast cells")) #  Subset-col, cells_of_interest
subsetting <- c(NULL, NULL) #  Subset-col, cells_of_interest

norm_meth_scale <- c(TRUE, "LogNormalize", 10000) # Perform log normalisatin, method, scale
vf <- c("vst", 2000, FALSE, FALSE) # VF method, amount, use old vf method, silence VDJgenes using scRep
scaling <- TRUE # w
var.to.regress <- NULL # supply the var to be regressed if valid
harmony <- T # Run harmony or not
n_PCS <- 50 # Amount of Pcs to include
clustering <- c(TRUE, 4) # Perform clustering, Algorithm
sub_dir <- NULL


feature_vec_chosen <- feature_vec_overall
```

```{r}
seurat_obj_path <- ifelse(!is.null(sub_dir), here(obj_path, sub_dir, paste0(Processed_name, ".qs2")), here(obj_path, paste0(Processed_name, ".qs2")))
seurat_obj <- qs_read(seurat_obj_path)
.size(seurat_obj)
#https://www.nature.com/articles/s41556-025-01840-5

# seurat_obj@meta.data  <- seurat_obj@meta.data %>%
#     mutate(Ann_lvl_2 = recode(Ann_lvl_2, `Ma-Inflam` = 'Ma_SLC2A1+' ))
# seurat_obj@meta.data$Ann_lvl_2 %>% table()
# qs_save(seurat_obj, seurat_obj_path)
```

```{r, eval = F}
meta_df <- seurat_obj@meta.data
meta_df[,c(9:30,38:40)] %>% distinct() -> new_df
new_df %>% tibble::remove_rownames() -> distinct_meta
check <- lapply(colnames(distinct_meta),\(c){if(class(distinct_meta[[c]]) == 'list'){distinct_meta[[c]] <- as.character(distinct_meta[[c]])} else {distinct_meta[[c]]} }) %>% as.data.frame()
colnames(check) <- colnames(distinct_meta)
write.csv(check, here('Glioma_Project/01_data/00_metadata.csv'))

seurat_obj@meta.data$file_name %>% unique()
(seurat_obj$cell_het_paper_id == 'ODG-7') %>% table() # Strange
```


```{r cleaning the objects}
p <- VlnPlot(seurat_obj, features = c('nCount_RNA','nFeature_RNA','percent_mt'), group.by = 'Ann_lvl_2', ncol = 2) 
# pdf(here('Glioma_Project/04_figs/00Test.pdf'), width= 15, height = 20)
# p
# dev.off()
Idents(seurat_obj) <- 'Ann_lvl_2'
seurat_obj$log_nCount_RNA <- log(seurat_obj$nCount_RNA, base = 10)
seurat_obj$log_nFeature_RNA <- log(seurat_obj$nFeature_RNA, base = 10)
Myeloid_anns <- c("Mg_classic", "Mg_Inflam", "Mg_Anti-Inflam", "Mg_PLCG2", "Mg_active", "BAMs", "TAMs", "Ma_SLC2A1+", "Ma-IFN", "Ma-Mt+", "Mono-Classic","Myeloid_Prolif","DCs_CD1C+","pDCs",'Mast-cells')
T_anns <- c('CD8+_TNFSF9','CD8+_Cytotoxic','CD8+ Eff-Exh','CD4+ cm','CD4+ Tfh','Tregs','T-IFN','T_Prolif','MAIT-cells','NK')
seurat_obj_M <- subset(x = seurat_obj, idents = Myeloid_anns, invert = F)
seurat_obj_T <- subset(x = seurat_obj, idents = T_anns, invert =F)
```

```{r T cells}
obj <- seurat_obj_T
reduction_name <- 'T'
samples_to_use <- obj@meta.data %>% group_by(sample_name) %>% summarise(count = n()) %>% filter(count> 1) %>% select(sample_name) %>%unlist(as.vector(.))# as.vector()
obj <- subset(obj, subset = sample_name %in% samples_to_use)
obj[["RNA"]] <- split(obj[["RNA"]], f = obj@meta.data[["sample_name"]])
obj <- FindVariableFeatures(obj, selection.method = 'vst', nfeatures = 1000, verbose = F)
VariableFeatures(obj) <- .remove_VDJ(VariableFeatures(obj))
obj <- ScaleData(object = obj, features = VariableFeatures(obj), verbose = FALSE)
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = 100
    )
ElbowPlot(obj, reduction = paste0("pca_", reduction_name), ndims = 100)
VizDimLoadings(obj, dims = 1:5, reduction = paste0("pca_", reduction_name), nfeatures = 15)
n_PCS <- 20
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = n_PCS
    )

obj <- IntegrateLayers(
            object = obj, method = HarmonyIntegration,
            orig.reduction = paste0("pca_", reduction_name), new.reduction = paste0("harmony_pca.", reduction_name),
            verbose = TRUE
        )
obj <- FindNeighbors(obj, reduction = paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS))
res_values <- c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5, 0.7, 0.9, 1.5)
obj <- FindClusters(obj, resolution = res_values, algorithm = clustering[2])
obj <- RunUMAP(obj, reduction =  paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS), reduction.name =  paste0("harmony_umap.", reduction_name))
obj <- JoinLayers(obj)



p1 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation T cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
resolution_columns <- grep("^RNA_snn_res\\.",
    colnames(obj@meta.data),
    value = TRUE
)
p2 <- .seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name), group = resolution_columns, label = TRUE) & NoLegend()
p1 + p2
Idents(obj) <- 'Ann_lvl_2'
marker_list <- FindAllMarkers(
    object = obj,
    only.pos = F,
    min.pct = 0.25,
    logfc.threshold = 0.25
)
.marker_plot <- \(obj,marker_list,nmarkers = 5,cluster_var = F){
    list_genes_clust <-marker_list %>%
        arrange(cluster, desc(avg_log2FC)) %>% # Arrange within each cluster
        group_by(cluster) %>%
        select(cluster,gene) %>%
        slice_head(n = nmarkers) %>%
        group_split() %>% # Split into list by 'cluster'
        setNames(unique(markers$cluster)) 
    features <- unique(unlist(lapply(list_genes_clust, \(df){df$gene}),use.names = F))
    p <- DotPlot(obj, features = features, cluster.idents = cluster_var) + theme_bw() + RotatedAxis()
    return(p)
}

.marker_plot(obj,marker_list)
ggsave(here('Glioma_Project/04_figs/markers_T_dot.pdf'), width = 12, height = 4.5)
p1 <- VlnPlot(obj, features = c('log_nCount_RNA','log_nFeature_RNA','percent_mt'), group.by = 'Ann_lvl_2', ncol = 2) 
p2 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation T cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
p3 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = feature_vec_T) & NoLegend()
plot <- p1 + p2
ggsave(here('Glioma_Project/04_figs/violin_T.pdf'), width = 12, height = 10, plot = plot)
ggsave(here('Glioma_Project/04_figs/markers_T.pdf'), width = 15, height = 30, plot = p3)

qs_save(obj,here('Glioma_Project/03_seurat_objects/Processed_T.qs2'))
```

```{r myeloid}
obj <- seurat_obj_M
reduction_name <- 'M'
samples_to_use <- obj@meta.data %>% group_by(sample_name) %>% summarise(count = n()) %>% filter(count> 1) %>% select(sample_name) %>%unlist(as.vector(.))# as.vector()
obj <- subset(obj, subset = sample_name %in% samples_to_use)
obj[["RNA"]] <- split(obj[["RNA"]], f = obj@meta.data[["sample_name"]])
obj <- FindVariableFeatures(obj, selection.method = 'vst', nfeatures = 1000, verbose = F)
VariableFeatures(obj) <- .remove_VDJ(VariableFeatures(obj))
obj <- ScaleData(object = obj, features = VariableFeatures(obj), verbose = FALSE)
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = 100
    )
ElbowPlot(obj, reduction = paste0("pca_", reduction_name), ndims = 100)
VizDimLoadings(obj, dims = 1:5, reduction = paste0("pca_", reduction_name), nfeatures = 15)
n_PCS <- 20
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = n_PCS
    )

obj <- IntegrateLayers(
            object = obj, method = HarmonyIntegration,
            orig.reduction = paste0("pca_", reduction_name), new.reduction = paste0("harmony_pca.", reduction_name),
            verbose = TRUE
        )
obj <- FindNeighbors(obj, reduction = paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS))
res_values <- c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5, 0.7, 0.9, 1.5)
obj <- FindClusters(obj, resolution = res_values, algorithm = clustering[2])
obj <- RunUMAP(obj, reduction =  paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS), reduction.name =  paste0("harmony_umap.", reduction_name))
obj <- JoinLayers(obj)

p1 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation T cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
resolution_columns <- grep("^RNA_snn_res\\.",
    colnames(obj@meta.data),
    value = TRUE
)
p2 <- .seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name), group = resolution_columns, label = TRUE) & NoLegend()
p1 + p2 # quick vis


# Idents(obj) <- 'Ann_lvl_2'
# marker_list <- FindAllMarkers(
#     object = obj,
#     only.pos = F,
#     min.pct = 0.25,
#     logfc.threshold = 0.25
# )

# .marker_plot(obj,marker_list)
# ggsave(here('Glioma_Project/04_figs/markers_M_dot.pdf'), width = 15, height = 4.5)
# p1 <- VlnPlot(obj, features = c('log_nCount_RNA','log_nFeature_RNA','percent_mt'), group.by = 'Ann_lvl_2', ncol = 2) 
# p2 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation T cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
# p3 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = Mgenes) & NoLegend()
# plot <- p1 + p2
# ggsave(here('Glioma_Project/04_figs/violin_M.pdf'), width = 12, height = 10, plot = plot)
# ggsave(here('Glioma_Project/04_figs/markers_M.pdf'), width = 15, height = 30, plot = p3)
qs_save(obj,here('Glioma_Project/03_seurat_objects/Processed_M.qs2'))

```



```{r subset M}
seurat_obj <- qs_read(here(paste0('Glioma_Project/03_seurat_objects/Processed_M.qs2')))
seurat_obj_subset <- subset(seurat_obj, subset = Ann_lvl_1H %in% c('Microglia','Macrophage','Monocyte','Myeloid_Prolif'))
seurat_obj_subset <- AddModuleScore_UCell(
    seurat_obj_subset,
    features = gene_list)
reduction_name <- 'M_M'
obj <- seurat_obj_subset
samples_to_use <- obj@meta.data %>% group_by(sample_name) %>% summarise(count = n()) %>% filter(count> 1) %>% select(sample_name) %>%unlist(as.vector(.))# as.vector()
obj <- subset(obj, subset = sample_name %in% samples_to_use)
obj[["RNA"]] <- split(obj[["RNA"]], f = obj@meta.data[["sample_name"]])
obj <- FindVariableFeatures(obj, selection.method = 'vst', nfeatures = 1000, verbose = F)
VariableFeatures(obj) <- .remove_VDJ(VariableFeatures(obj))
obj <- ScaleData(object = obj, features = VariableFeatures(obj), verbose = FALSE)
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = 100
    )
ElbowPlot(obj, reduction = paste0("pca_", reduction_name), ndims = 100)
VizDimLoadings(obj, dims = 1:5, reduction = paste0("pca_", reduction_name), nfeatures = 15)
n_PCS <- 20
obj <- RunPCA(
        object = obj, features = VariableFeatures(obj),
        nfeatures.print = 5, ndims.print = 1:2,
        reduction.name = paste0("pca_", reduction_name),
        npcs = n_PCS
    )

obj <- IntegrateLayers(
            object = obj, method = HarmonyIntegration,
            orig.reduction = paste0("pca_", reduction_name), new.reduction = paste0("harmony_pca.", reduction_name),
            verbose = TRUE
        )
obj <- FindNeighbors(obj, reduction = paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS))
res_values <- c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5, 0.7, 0.9, 1.5)
obj <- FindClusters(obj, resolution = res_values, algorithm = clustering[2])
obj <- RunUMAP(obj, reduction =  paste0("harmony_pca.", reduction_name), dims = 1:as.numeric(n_PCS), reduction.name =  paste0("harmony_umap.", reduction_name))
obj <- JoinLayers(obj)

p1 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation T cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
resolution_columns <- grep("^RNA_snn_res\\.",
    colnames(obj@meta.data),
    value = TRUE
)
qs_save(obj,here(paste0('Glioma_Project/03_seurat_objects/Processed_',reduction_name,'.qs2')))

Idents(obj) <- 'Ann_lvl_2'
marker_list <- FindAllMarkers(
    object = obj,
    only.pos = F,
    min.pct = 0.25,
    logfc.threshold = 0.25
)
p <- .marker_plot(obj,marker_list)
p1 <- VlnPlot(obj, features = c('log_nCount_RNA','log_nFeature_RNA','percent_mt'), group.by = 'Ann_lvl_2', ncol = 2) 
p2 <-.seurat_dimplot(obj, reduction = paste0("harmony_umap.", reduction_name),label = T, title = 'Annotation M_M cells',group = 'Ann_lvl_2', pt.size = 1) #+ theme(legend.position = 'none')
p3 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = Mgenes) & NoLegend()
p4 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = 'SLC2A1') & NoLegend()
plot <- p1 + p2
ggsave(here('Glioma_Project/04_figs/markers_M_M_dot.pdf'),plot = p, width = 15, height = 4.5)
ggsave(here('Glioma_Project/04_figs/violin_M_M.pdf'), width = 12, height = 10, plot = plot)
ggsave(here('Glioma_Project/04_figs/markers_M_M.pdf'), width = 15, height = 30, plot = p3)
pdf(here('Glioma_Project/04_figs',paste0('Overview_figures',reduction_name),'.pdf'))
p
plot
p3
p4 + p2
dev.off()

# M1 <- c('FCGR2B','FCGR1A','CD68','CD80','CD86' ) #CD32, CD64, CD68, CD80, and CD86, 
# feature_vec <- c('IL1A','IL1B','IL6','IL4',"TNF",'CCL2',"CCL3","CCL5",'CXCL1',"CXCL2","IFNG")
# p5 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = feature_vec) & NoLegend()
# p6 <- .feature_plot(obj, reduction = paste0("harmony_umap.", reduction_name), features = M1) & NoLegend()

```